src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateArray = [];
+  let newState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+
+        break;
+
+      default:
+        break;
+    }
+    stateArray.push({ ...newState });
```

Comment:
Instead of creating a new copy of the state inside the loop after each action, you can push the copy to the `stateArray` outside of the `switch` block to avoid repetition. This will make the code cleaner and more efficient.